### PR TITLE
Fix overflow when `ctx.host_current_height().revision_height < ctx.host_chain_history_size()`

### PR DESCRIPTION
**Pull request by vitorenesduarte**
Closes: #685 

## Description

______

For contributor use:

- [ ] Updated the __Unreleased__ section of [CHANGELOG.md](https://github.com/informalsystems/ibc-rs/blob/master/CHANGELOG.md) with the issue.
- [ ] If applicable: Unit tests written, added test to CI.
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Updated relevant documentation (`docs/`) and code comments.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.

### CODE REVIEW

File: modules/src/ics03_connection/handler/verify.rs [hunk 1]
```diff
@@ -197,10 +197,7 @@ pub fn check_client_consensus_height(
         return Err(Kind::InvalidConsensusHeight(claimed_height, ctx.host_current_height()).into());
     }
 
-    let oldest_available_height =
-        ctx.host_current_height().revision_height - ctx.host_chain_history_size() as u64;
```
Comment (adizere) [inline modules/src/ics03_connection/handler/verify.rs:201]: Can we remove method `host_chain_history_size`, which is no longer useful now that you introduced `host_oldest_height`?

File: CHANGELOG.md [hunk 2]
```diff
@@ -32,11 +32,10 @@
...
 [comment]: <> (  - [nothing yet])
 
-[comment]: <> (### BUG FIXES)
-
-[comment]: <> (- [ibc])
+### BUG FIXES
 
-[comment]: <> (  - [nothing yet])
+- [ibc]
+  - Overflow if `ctx.host_current_height().revision_height < ctx.host_chain_history_size()` ([#685])
```
Comment (adizere) [inline CHANGELOG.md:38]: Nit: Would be nice to have a slightly higher level description in the changelog.

```suggestion
  - Fix overflow bug in ICS03 client consensus height verification method ([#685])
```

**Review by adizere — CHANGES_REQUESTED** [review]
Nice find, Vitor! I had two minor remarks.

File: modules/src/ics03_connection/handler/verify.rs (hunk 1 above)
Comment (vitorenesduarte) [reply modules/src/ics03_connection/handler/verify.rs:201]: Done in c851c336c039286ef803cdae5e693d058cf0c342.

**Comment by vitorenesduarte** [comment]
Thanks for the review @adizere. I've addressed your suggestions.

I'm thinking that the tests for connection open ack were relying on wrong behaviour since they started failing with this PR. I tried to fix them in 4da1d34c9d0505a360060c69957394b9bb55690a but there's one (now commented out) I didn't figure out how.

**Comment by vitorenesduarte** [comment]
> I'm thinking that the tests for connection open ack were relying on wrong behaviour since they started failing with this PR. I tried to fix them in [4da1d34](https://github.com/informalsystems/ibc-rs/commit/4da1d34c9d0505a360060c69957394b9bb55690a) but there's one (now commented out) I didn't figure out how.

The test that was commented out is expecting a `MissingLocalConsensusState` to be thrown here: https://github.com/informalsystems/ibc-rs/blob/vitor/overflow/modules/src/ics03_connection/handler/verify.rs#L167-L170

However, before this, we're already checking if the height is stale here: https://github.com/informalsystems/ibc-rs/blob/vitor/overflow/modules/src/ics03_connection/handler/verify.rs#L200-L203

If the height is not stale, then the consensus state must exist, in which case the error `MissingLocalConsensusState` should instead be a panic. If you agree with this analysis, I'll replace the error with a panic in a follow-up PR.

**Review by adizere — APPROVED** [review]
> I'm thinking that the tests for connection open ack were relying on wrong behaviour since they started failing with this PR. I tried to fix them in 4da1d34 but there's one (now commented out) I didn't figure out how.

That may well be. Tried to consolidate tests in that module recently, but there's still some work to be done. Your improvements definitely help!

**Comment by adizere** [comment]
> > I'm thinking that the tests for connection open ack were relying on wrong behaviour since they started failing with this PR. I tried to fix them in [4da1d34](https://github.com/informalsystems/ibc-rs/commit/4da1d34c9d0505a360060c69957394b9bb55690a) but there's one (now commented out) I didn't figure out how.
> 
> The test that was commented out is expecting a `MissingLocalConsensusState` to be thrown here: https://github.com/informalsystems/ibc-rs/blob/vitor/overflow/modules/src/ics03_connection/handler/verify.rs#L167-L170
> 
> However, before this, we're already checking if the height is stale here: https://github.com/informalsystems/ibc-rs/blob/vitor/overflow/modules/src/ics03_connection/handler/verify.rs#L200-L203
> 
> If the height is not stale, then the consensus state must exist, in which case the error `MissingLocalConsensusState` should instead be a panic. If you agree with this analysis, I'll replace the error with a panic in a follow-up PR.

I agree with your analysis. To summarize: in conn open try & ack handlers there are two verifications on the field `msg.proofs.consensus_proof().height` as follows:

1. first we verify for staleness with [`check_client_consensus_height`](https://github.com/informalsystems/ibc-rs/blob/vitor/overflow/modules/src/ics03_connection/handler/verify.rs#L200-L203), i.e., verify that this height is not older than the oldest height of the host (local) chain, nor newer than the current height of the host chain.
2. second we verify that the consensus state for this height exists in the local chain, as part of [`verify_consensus_proof`](https://github.com/informalsystems/ibc-rs/blob/vitor/overflow/modules/src/ics03_connection/handler/verify.rs#L167-L170).

I think the first verification is more of a fail-fast. While the second verification is a pre-check as part of the whole ensemble of proof verification. I agree there is some redundancy there. I would suggest to open an issue documenting this, but I wouldn't fix things with a panic (we have no panics in the modules yet, and not sure if it's a good idea to fix it like that).

**Comment by vitorenesduarte** [comment]
I've updated #698. @adizere please confirm that it reflects the discussion here. 

Regarding `panic`s, there are already some `unwrap`s, which can also panic. My view is that both are okay as long as they are prefixed with some explanation why there are "safe".

**Pull request merged**